### PR TITLE
Assume unresolved packaging variables in a POM mean jar

### PR DESCRIPTION
**Pull request by big-guy**
Maven ignores the declared packaging and assumes jar unless told otherwise.

Gradle attempts to resolve a dependency's artifact based on its packaging.

In the past, we would have attempted to resolve an artifact with the extension ${packaging.type} and, failing that, tried jar.

Now, when we detect that the packaging isn't something valid, we immediately default to jar in our internal representation of the POM.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/PomReader.java
Comment (jvandort) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/PomReader.java:744]: Is this necessarily true? Can you have a `$` that is not a pattern? Can we expand this to a regex that checks for `$\{.*\}`?

**Review by jvandort — APPROVED** [review]

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/PomReader.java
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/PomReader.java:742]: 🤔 Should we warn in this case?  This possible mistake remains completely invisible now.  Clearly attributing the problem to the particular dependency with the bad POM would (eventually) bring it to the attention of that lib's maintainers, who could correct their build in the next version.

I suppose if we warn we'd also need a flag to disable this warning (perhaps for specific POMs) so people could keep their build output clean.

File: platforms/software/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/PomReaderTest.groovy
Comment (tresat) [inline platforms/software/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/PomReaderTest.groovy:177]: 💭 Related to the lack of a warning - wouldn't both this test (and the new integration test) pass prior to this change?  This doesn't have a way of verifying the changed behavior.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/PomReader.java
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/PomReader.java:346]: 💅 Because of confusion with Kotlin, Lombok and JEPs, and because it is carries little meaning, I think `val` shouldn't be used as a variable name.  Why not use just `packaging`?

**Review by tresat — APPROVED** [review]
Seems to me that it's worth considering a warning/some way of notifying the user about this situation, but not worth holding this up.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/PomReader.java
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/PomReader.java:742]: I can see why we would want to warn, but it also feels odd to me to warn when wee are following maven's default behavior.

If maven does this and does not warn, I don't think we should warn

File: platforms/software/dependency-management/src/main/java/org/gradle/internal/component/external/model/maven/RealisedMavenModuleResolveMetadata.java
Comment (jvandort) [inline platforms/software/dependency-management/src/main/java/org/gradle/internal/component/external/model/maven/RealisedMavenModuleResolveMetadata.java:217]: Why is this static method located on the `PomReader` if it is not used there? It seems like it would be better located in this class

Comment (big-guy) [reply platforms/software/dependency-management/src/main/java/org/gradle/internal/component/external/model/maven/RealisedMavenModuleResolveMetadata.java:217]: I was trying to keep the knowledge about what a valid Maven substitution is somewhere close to where that substitution happens. The pattern used by hasUnresolvedSubstitutions is from IvyPatternHelper and that's extensively used by PomReader.

The "better" fix would be to bring IvyPatternHelper into gradle/gradle like we did with PomReader and move hasUnresolvedSubstitutions into that.

**Review by jvandort — APPROVED** [review]

**Comment by big-guy** [comment]
@bot-gradle test and merge

**Pull request merged**
